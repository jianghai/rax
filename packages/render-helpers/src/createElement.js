import Host from './host';
import createEmptyElement from './createEmptyElement';

/**
 * Create MiniApp element.
 * @param type {String|Object} element type.
 * @param props {Object} element props.
 * @param children {Element|Array<Element>} child elements.
 * @return {Element} element.
 */
export default function(type, props, ...children) {
  if (typeof type === 'object' && type.__esModule) {
    type = type.default;
  } else if (type === '$template') {
    if (props.is) {
      return props.is.call(props.pageInstance, props.data);
    } else {
      return createEmptyElement();
    }
  }

  // For reduce code size
  // eg: _c(View, undefined, ['xxx'])
  // vs: _c(View, ['xxx'])
  if (Array.isArray(props)) {
    children = props;
    props = undefined;
  }

  return Host.render.createElement(type, props, ...children);
}


