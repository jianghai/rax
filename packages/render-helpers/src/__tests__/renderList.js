import * as Rax from 'rax';
import setRender from '../setRender';
import renderList from '../renderList';
import createElement from '../createElement';

setRender(Rax);
function render(value, index) {
  return createElement('view', [value, index]);
}

describe('renderList', () => {
  it('should render list by number', () => {
    const el = renderList(5, render);
    expect(el).toMatchSnapshot();
  });

  it('should render list by string', () => {
    const el = renderList('hello world', render);
    expect(el).toMatchSnapshot();
  });

  it('should render list by array', () => {
    const el = renderList([0, 1, 2, 3, 4], render);
    expect(el).toMatchSnapshot();
  });

  it('should render list by object', () => {
    const el = renderList({
      a: 1,
      b: 2,
      c: 3,
    }, render);
    expect(el).toMatchSnapshot();
  });
});
