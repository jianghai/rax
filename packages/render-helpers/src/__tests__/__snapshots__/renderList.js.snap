// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderList should render list by array 1`] = `
Array [
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        0,
        0,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        1,
        1,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        2,
        2,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        3,
        3,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        4,
        4,
      ],
    },
    "ref": null,
    "type": "view",
  },
]
`;

exports[`renderList should render list by number 1`] = `
Array [
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        1,
        0,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        2,
        1,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        3,
        2,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        4,
        3,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        5,
        4,
      ],
    },
    "ref": null,
    "type": "view",
  },
]
`;

exports[`renderList should render list by object 1`] = `
Array [
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        1,
        "a",
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        2,
        "b",
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        3,
        "c",
      ],
    },
    "ref": null,
    "type": "view",
  },
]
`;

exports[`renderList should render list by string 1`] = `
Array [
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        "h",
        0,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        "e",
        1,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        "l",
        2,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        "l",
        3,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        "o",
        4,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        " ",
        5,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        "w",
        6,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        "o",
        7,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        "r",
        8,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        "l",
        9,
      ],
    },
    "ref": null,
    "type": "view",
  },
  Object {
    "_owner": null,
    "key": null,
    "props": Object {
      "children": Array [
        "d",
        10,
      ],
    },
    "ref": null,
    "type": "view",
  },
]
`;
